File: day-11-review-day-1/python/linked-list.py
# Create node
def create_node(value):
  return { "value": value, "next": None }

# Pre append an head to the beggining
def prepend(head, value):
  new_node = create_node(value)
  new_node["next"] = head
  return new_node

# Prints all values from head to end
def traverse(head):
  current = head
  while current is not None:
    print(current["value"])
    current = current["next"]

# Find node with given value, return node or null 
def find(head, value):
  current = head
  while current is not None:
    if current["value"] == value:
      return current
    current = current["next"]
  return None

#  Insert a new value after the given node
def insert_after(node, value):
  new_node = create_node(value)
  new_node["next"] = node["next"]
  node["next"] = new_node

# Test
head = create_node(10)
print(head)
print("\n create node ===================\n")

head = prepend(head, 5)
print(head)
print("\n prepend 5 ===================\n")

head = prepend(head, 3)
print(head)
print("\n prepend 3 ===================\n")

traverse(head)
print("\n traverse ===================\n")

found = find(head, 5)
if found:
    insert_after(found, 7)
print(head)
print("\n find and insert after ===========\n")

traverse(head)

File: day-11-review-day-1/js/linked-list.js
// Create node
function createNode(value) {
  return { value: value, next: null };
}

// Pre append an head to the beggining
function prepend(head, value) {
  let newNode = createNode(value);
  newNode.next = head;
  return newNode;
}

// Prints all values from head to end
function traverse(head) {
  let current = head;
  while (current !== null) {
    console.log(current.value);
    current = current.next;
  }
}

// Find node with given value,  return node or null
function find(head, value) {
  let current = head;
  while (current !== null) {
    if (current.value === value) {
      return current;
    }
    current = current.next;
  }
  return null;
}

// Insert a new value after the given node
function insertAfter(node, value) {
  let newNode = createNode(value);
  newNode.next = node.next;
  node.next = newNode;
}

let head = createNode(5);

head = prepend(head, 15);

// console.log(find(head, 5));
insertAfter(head, 10)
let found = find(head, 10)
console.log("--->", found);
if (found) {
  insertAfter(found, 4)
}
console.log(head);
